### PR TITLE
8276743: Make openjdk build Zip Archive generation "reproducible"

### PR DESCRIPTION
**Pull request by andrew-m-leonard**
This PR adds a new openjdk build tool MakeZipReproducible, which if ENABLE_REPRODUCIBLE_BUILD is enabled, generates a final "zip" file in a deterministic way, ensuring ordering and timestamps are set as specified.

Using this tool in ZipArchive.gmk will ensure src.zip is then created deterministically.

Signed-off-by: Andrew Leonard <anleonar@redhat.com>

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8276743](https://bugs.openjdk.java.net/browse/JDK-8276743): Make openjdk build Zip Archive generation "reproducible"

### Reviewers
 * [Erik Joelsson](https://openjdk.java.net/census#erikj) (@erikj79 - **Reviewer**)
 * [Magnus Ihse Bursie](https://openjdk.java.net/census#ihse) (@magicus - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6311/head:pull/6311` \
`$ git checkout pull/6311`

Update a local copy of the PR: \
`$ git checkout pull/6311` \
`$ git pull https://git.openjdk.java.net/jdk pull/6311/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6311`

View PR using the GUI difftool: \
`$ git pr show -t 6311`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6311.diff">https://git.openjdk.java.net/jdk/pull/6311.diff</a>

</details>

### CODE REVIEW

File: make/common/ZipArchive.gmk [hunk 1]
```diff
@@ -26,6 +26,8 @@
 ifndef _ZIP_ARCHIVE_GMK
 _ZIP_ARCHIVE_GMK := 1
 
+include ../ToolsJdk.gmk
```
Comment (erikj79) [inline make/common/ZipArchive.gmk:29]: Should probably add a comment about inclusion of this file now requires an explicit dependency on build-tools in Main.gmk.

File: make/jdk/src/classes/build/tools/generatezip/GenerateZip.java [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+
+        return true;
+    }
+
+    // Walk tree matching files and adding to entries list
+    void processFiles(Path path) throws IOException {
+        File fpath = path.toFile();
+        boolean pathIsDir = fpath.isDirectory();
+
+        // Keep a sorted Set of files to be processed, so that the Jmod is reproducible
```
Comment (erikj79) [inline make/jdk/src/classes/build/tools/generatezip/GenerateZip.java:161]: Aren't we generating zip files rather than jmod files here?

**Review by erikj79 — COMMENTED** [review]
I think it would it make sense add a configure parameter to to enable/disable this new functionality. I think developers would want to opt out of extra unnecessary build steps when building and testing locally. Not sure what the default should be. For other reproducible measures, we require active enabling today.

File: make/common/ZipArchive.gmk [hunk 3]
```diff
@@ -156,14 +160,24 @@ define SetupZipArchiveBody
...
 	        $$($1_ZIP_EXCLUDES_$$s) \
 	        || test "$$$$?" = "12" \
 	    ))$$(NEWLINE) \
-	) true \
+	) true
+	$$(call ExecuteWithLog, \
+	    $$(SUPPORT_OUTPUTDIR)/generatezip/$$(patsubst $$(OUTPUTDIR)/%,%, $$@), \
+	    (cd $$(SUPPORT_OUTPUTDIR)/ziptmp/$1/files && \
+	     $(RM) $$@ && \
+	     $(UNZIP) -q $$(SUPPORT_OUTPUTDIR)/ziptmp/$1/tmp.zip && \
```
[the remark this answers is not in the document]
Comment (andrew-m-leonard) [reply make/common/ZipArchive.gmk:178]: @erikj79 sure, I can look at doing that

File: make/jdk/src/classes/build/tools/generatezip/GenerateZip.java (hunk 2 above)
Comment (andrew-m-leonard) [reply make/jdk/src/classes/build/tools/generatezip/GenerateZip.java:161]: copy&paste error...

**Comment by andrew-m-leonard** [comment]
> I think it would it make sense add a configure parameter to to enable/disable this new functionality. I think developers would want to opt out of extra unnecessary build steps when building and testing locally. Not sure what the default should be. For other reproducible measures, we require active enabling today. Maybe default on for release builds and off for debug builds.

@erikj79 so I was hoping to not add another configure arg, when the intentions of this should just all be positive, ordered&deterministic. Also @magicus I get the impression from your comments that we don't really want to end up with in order to build reproducibly you have to specify all these options x,y,z,...
Maybe we should just have one big flag --with-reproducible-build ?

**Comment by magicus** [comment]
We already have an --enable-reproducible-builds. If (and I say if) we need to turn this on/off with a flag, this would to fine.

However, as I have said previously in a private discussion with Andrew, I prefer it if we can make reproducible builds so cheap, reliable and uncontroversial so we can always to reproducible builds, and remove that flag.

For this case, I think it depends on two things:
1) the extra time to make the zip file reproducible. Some benchmarking on the GenerateZip for src.zip would be good to have, at least for some ballpark figures.

2) When is src.zip built? (I don't remember right now) If it is part of the exploded-image, then it is really time sensitive (and should maybe move out of that target). If it part of the jdk-image, I'd say it is not as sensitive, due to
a) this is very much slower anyway, and
b) that part is much more parallelizible, so src.zip can be produced while waiting for jlink or whatever.

**Comment by magicus** [comment]
Also, the name is not really a correct description at this point. Maybe a straight `MakeZipReproducible` instead or something like that?

**Comment by magicus** [comment]
And have you verified that running with `zip -X` is not enough? I remember checking into this before, but I don't remember the conclusions. The purpose is to leave out some extra information, such as time stamps, but it might not be enough to guarantee reproducible builds.

**Comment by andrew-m-leonard** [comment]
> And have you verified that running with `zip -X` is not enough? I remember checking into this before, but I don't remember the conclusions. The purpose is to leave out some extra information, such as time stamps, but it might not be enough to guarantee reproducible builds.

Correct, I have tried that already, that does remove the extended timestamp information, and as long as the input source is already correctly timestamped it works to a point, that point being the "file ordering" is not-deterministic and I think is dependent on how the OS file queries returns files. From history this is what I tried: https://github.com/andrew-m-leonard/jdk-1/blob/f18b5826b9ae3e7e5750190a669edcfd8bb8b2cc/make/common/ZipArchive.gmk#L166

**Comment by andrew-m-leonard** [comment]
> We already have an --enable-reproducible-builds. If (and I say if) we need to turn this on/off with a flag, this would to fine.
> 
> However, as I have said previously in a private discussion with Andrew, I prefer it if we can make reproducible builds so cheap, reliable and uncontroversial so we can always to reproducible builds, and remove that flag.
> 
> For this case, I think it depends on two things:
> 
>     1. the extra time to make the zip file reproducible. Some benchmarking on the GenerateZip for src.zip would be good to have, at least for some ballpark figures.
> 
>     2. When is src.zip built? (I don't remember right now) If it is part of the exploded-image, then it is really time sensitive (and should maybe move out of that target). If it part of the jdk-image, I'd say it is not as sensitive, due to
>        a) this is very much slower anyway, and
>        b) that part is much more parallelizible, so src.zip can be produced while waiting for jlink or whatever.

@magicus
1) I'll do some rough benchmarks
2) "zip-source" is only dependent on "gensrc", ie.all module -gensrc stages must have completed, so in theory it should run in parallel while the linking goes on, a quick scan of a parallel build log seems to confirm that, the "Updating support/src.zip" occurs quite some time before the exploded image linking:
```
20:34:45  Compiling 31 files for jdk.management.agent
20:34:45  Compiling 16 files for jdk.security.jgss
20:34:45  Compiling 30 files for jdk.security.auth
20:34:45  Creating support/native/java.security.jgss/libj2gss/static/libj2gss.a from 3 file(s)
20:34:45  Updating support/src.zip
20:34:46  Creating support/native/jdk.management.agent/libmanagement_agent/static/libmanagement_agent.a from 1 file(s)
20:34:46  Creating support/native/jdk.security.auth/libjaas/static/libjaas.a from 1 file(s)
20:34:48  Compiling 136 files for jdk.jdeps
20:34:48  Creating support/test/lib-test/jtreg/native/bin/jvm-test-launcher from 1 file(s)
20:34:49  Creating support/modules_libs/java.base/libverify.so from 1 file(s)
.........
20:36:17  Optimizing the exploded image
20:36:18  Creating java.datatransfer.jmod
20:36:18  Creating java.instrument.jmod
20:36:18  Creating java.compiler.jmod
```

File: make/common/ZipArchive.gmk (hunk 3 above)
[the remark this answers is not in the document]
Comment (erikj79) [reply make/common/ZipArchive.gmk:178]: You are already keeping all the filenames in memory for sorting, so reading up the ZipEntry:s isn't that much more data, just some extra metadata for each entry. The actual file contents is not part of the ZipEntry object. When actually copying the files, you can use the ZipFile class to access ZipEntry's in arbitrary order to read their streams as InputStream.

**Comment by erikj79** [comment]
I agree that ideally reproducibility should be on by default, but if there is a cost, then you can be sure OpenJDK developers will be looking for a way to remove it for faster turnaround times. I would propose a specific configure parameter for this specific case, reproducible zip files, that is default on for release builds and off for debug builds (debug builds aren't reproducible by nature) and let the existing meta flag also control the value of this new flag.

File: make/common/ZipArchive.gmk (hunk 3 above)
[the remark this answers is not in the document]
Comment (erikj79) [reply make/common/ZipArchive.gmk:178]: Actually, you don't even need to save the ZipEntry:s in memory, you can just extract filenames from them on the first pass, sort them, then lookup the entries in ZipFile again on the second lap. :) I don't think that's necessary though.

[the remark this answers is not in the document]
Comment (andrew-m-leonard) [reply make/common/ZipArchive.gmk:178]: @erikj79 thanks I didn't realize you can do that: "you can use the ZipFile class to access ZipEntry's in arbitrary order"

**Comment by magicus** [comment]
@erikj79 The flag --enable-reproducible-builds sets ENABLE_REPRODUCIBLE_BUILD in spec.gmk. This is set by our JIB profiles. I propose that we also turn it on for GHA builds. 

I think that the post-processing of the zip file can be dependent on this variable and that it serves no purpose to introduce a separate variable ENABLE_REPRODUCIBLE_ZIP that is set to the same value as ENABLE_REPRODUCIBLE_BUILD. Do you agree?

File: make/common/ZipArchive.gmk (hunk 1 above)
Comment (andrew-m-leonard) [reply make/common/ZipArchive.gmk:30]: done

File: make/common/ZipArchive.gmk (hunk 3 above)
[the remark this answers is not in the document]
Comment (andrew-m-leonard) [reply make/common/ZipArchive.gmk]: See new commit

**Comment by andrew-m-leonard** [comment]
@erikj79 @magicus I have just pushed a new commit with the suggested changes, and it works well, thank you for the help

I've also done a basic average benchmarking, on my rather slow Ubuntu VM:
- Existing src.zip processing : 18 seconds
- Additional MakeZipReproducible : 6 seconds

**Comment by magicus** [comment]
In the future, please refrain from force pushing to a PR. It makes history hard to follow for reviewers, and is generally strongly discouraged. OpenJDK uses the Skara tools which will automatically squash and rebase the commits in the PR.

**Comment by andrew-m-leonard** [comment]
> In the future, please refrain from force pushing to a PR. It makes history hard to follow for reviewers, and is generally strongly discouraged. OpenJDK uses the Skara tools which will automatically squash and rebase the commits in the PR.
@magicus I needed to cause a re-submit tests due to a macos timeout, and there seems no github Action or PR command to cause that, so I just force pushed, couldn't see any other way.... I presume there is?

**Comment by andrew-m-leonard** [comment]
@magicus is pushing an empty commit or dummy change preferable?

**Comment by andrew-m-leonard** [comment]
Looks like other PRs also getting mac or windows test failures..

**Comment by andrew-m-leonard** [comment]
@erikj79 all tests pass, ready for re-review please, thanks

**Comment by magicus** [comment]
@andrew-m-leonard Yes, pushing an empty commit is much better. The Skara tooling will automatically squash all commits in the PR when it is integrated, so it will be fully invisible in the end.

But there is a way to trigger re-runs, although I realize is has close to zero visibility (I'll need to blame Github for that :-().

Go to your personal fork of the JDK, select the "Actions" tab, select "Pre-submit tests" in the list to the left, and then press the `Run workflow` button in the cyan `This workflow has a workflow_dispatch event trigger.` field.

There you can select branch to run, and also additionally modify the set of platforms run.

File: make/Main.gmk [hunk 4]
```diff
@@ -508,13 +508,13 @@ $(eval $(call SetupTarget, docs-jdk-index, \
 $(eval $(call SetupTarget, docs-zip, \
     MAKEFILE := Docs, \
     TARGET := docs-zip, \
-    DEPS :=  docs-jdk, \
+    DEPS :=  docs-jdk buildtools-jdk, \
```
Comment (magicus) [inline make/Main.gmk:511]: Is this really needed? I did not think jrtfs-jar used zip?

**Comment by erikj79** [comment]
> @magicus is pushing an empty commit or dummy change preferable?

Yes, I think that's the only good way of re-triggering github actions. While working on a PR, it's better to let history be history so review comments don't get lost. When we finally integrate to mainline, everything will be squashed automatically.

File: make/Main.gmk (hunk 4 above)
Comment (magicus) [reply make/Main.gmk:511]: Sorry, I'm mis-reading the lines here. It's of course for docs-zip. Then it's perfectly in order! :-)

Comment (erikj79) [reply make/Main.gmk:511]: It's using JarArchive.gmk, which is a similar, but different and probably also needs the same treatment. We should probably share more code between them.

Comment (andrew-m-leonard) [reply make/Main.gmk:511]: I believe so, it's target docs-zip not jrtfs : https://github.com/openjdk/jdk/blob/51a5731d6dc4b6f6feac920a4b8b49c15fd6b34f/make/Docs.gmk#L712

**Review by magicus — APPROVED** [review]
I think it looks good now. Let Erik have a final say also, before integrating. 

Thank you for your effort!

**Review by erikj79 — APPROVED** [review]
Looks good.
